### ISSUE TITLE
ansible_lvm.lvs facts can't handle LVs with same name and different VGs

### ISSUE THREAD

**Issue by bryan-croft** [ansible/ansible#28679]
- Bug Report

##### COMPONENT NAME
LVM fact gathering

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = /home/guru/docker/ansible.cfg
  configured module search path = [u'/home/guru/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.4.0-py2.7.egg/ansible
  executable location = /bin/ansible
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]

```

##### CONFIGURATION
defaults

##### OS / ENVIRONMENT
N/A

##### SUMMARY
The hash produced by fact gathering for ansible_lvm uses the LV name as the key, but you can have multiple LV's on a system with the exact same name. In this case, only the last LV discovered is shown in the hash.

##### STEPS TO REPRODUCE
Create LVs with same name in different VGs. for example:
```
# lvs
  LV         VG  Attr       LSize   Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  demo       vg0 -wi-ao----   1.00g
  demo       vg1 -wi-a----- 248.00m
# ansible localhost -m setup
...snip...
"ansible_lvm": {
   "lvs": {
       "demo": {
           "size_g": "0.24",
           "vg": "vg1"
       }
```

##### EXPECTED RESULTS
Given data structure I guess the most I could hope for is an error. What I need is for both LVs to be listed even though they have the same name. Not sure how the LVM facts section could best be refactored to support this...

##### ACTUAL RESULTS
Only the last LV seen is shown, the other(s) become invisible.

**Comment by nrwahl2** [comment]
@bryan-croft AFAIK, there's no way to refactor the code in place without changing the data structures and breaking existing user code all over the place. I completely agree that the current output is not acceptable -- in my work environment we frequently have same-named lvols in different volgroups.

I lean toward creating a new item in the `ansible_lvm` fact that structures this information in a more logical, hierarchical manner without any components overwriting one another. This subfact could sit side-by-side with the existing `vgs/lvs/pvs` facts for a few minor releases before being deprecated in favor of the hierarchical model. We'll need to get further input from members of the core team.

This is my first pass. `lvm_tree` is the new subfact.
```
[reid@laptop ~/git]$ ansible localhost -b -c local -i hosts.ini -m setup -a 'filter=*lvm*'
localhost | SUCCESS => {
    "ansible_facts": {
        "ansible_lvm": {
            "lvm_tree": {
                "centos": {
                    "free_g": "0.01", 
                    "lvs": {
                        "home": {
                            "size_g": "1.00"
                        }, 
                        "root": {
                            "size_g": "27.05"
                        }, 
                        "swap": {
                            "size_g": "7.75"
                        }
                    }, 
                    "num_lvs": "3", 
                    "num_pvs": "1", 
                    "pvs": {
                        "/dev/sda9": {
                            "free_g": "0.01", 
                            "size_g": "35.80"
                        }
                    }, 
                    "size_g": "35.80"
                }, 
                "fedora": {
                    "free_g": "19.00", 
                    "lvs": {
                        "home": {
                            "size_g": "319.92"
                        }, 
                        "root": {
                            "size_g": "50.00"
                        }, 
                        "swap": {
                            "size_g": "7.75"
                        }
                    }, 
                    "num_lvs": "3", 
                    "num_pvs": "1", 
                    "pvs": {
                        "/dev/sda7": {
                            "free_g": "19.00", 
                            "size_g": "396.67"
                        }
                    }, 
                    "size_g": "396.67"
                }
            }, 
            "lvs": {
                "home": {
                    "size_g": "319.92", 
                    "vg": "fedora"
                }, 
                "root": {
                    "size_g": "50.00", 
                    "vg": "fedora"
                }, 
                "swap": {
                    "size_g": "7.75", 
                    "vg": "fedora"
                }
            }, 
            "pvs": {
                "/dev/sda7": {
                    "free_g": "19.00", 
                    "size_g": "396.67", 
                    "vg": "fedora"
                }, 
                "/dev/sda9": {
                    "free_g": "0.01", 
                    "size_g": "35.80", 
                    "vg": "centos"
                }
            }, 
            "vgs": {
                "centos": {
                    "free_g": "0.01", 
                    "num_lvs": "3", 
                    "num_pvs": "1", 
                    "size_g": "35.80"
                }, 
                "fedora": {
                    "free_g": "19.00", 
                    "num_lvs": "3", 
                    "num_pvs": "1", 
                    "size_g": "396.67"
                }
            }
        }
    }, 
    "changed": false, 
    "failed": false
}
```

**Comment by alikins** [comment]
Hmm, yeah, can't do much with the existing ansible_lvm data. 

If we add another fact with the info, we could try:

  - using 'lvm_full_name' for the key _if_ it is sufficiently unique
  - dropping the dict and returning a list/set of tuples that could have dupes on the lvs name
  - slightly more complicated... replacing the 1:1 key: lvs_data_structure with a 1:N. ie, making the
   right hand side a list of dicts instead of a single dict
  - (way more complicated) use/abuse json maps support for non unique keys, but that means replacing
    some of the simple python dicts used module side and controller side with a dict like class that supports 
    duplicate keys

Many other ways as well.

**Comment by nrwahl2** [comment]
Indeed! To me the tree model is more intuitive for LVM than a flat model -- lvols and pvols are <s>always</s> USUALLY components of volgroups. But, a flatter model like set of tuples or `lvm_full_name` could make access easier and be more intuitive to others.

I had to catch myself on the "always" when I realized a flaw in my fact: It is entirely possible to have a physical volume that is not yet part of a volume group. They all start out that way. No matter whether we go with a nested for flat solution, that will have to be addressed.
